src/Stratis.Bitcoin/AsyncWork/AsyncProvider.cs

```diff
@@ -0,0 +1,318 @@
+        /// <remarks>state can be either of type <see cref="IAsyncDelegateDequeuer{T}"/> or <see cref="IAsyncLoop"/></remarks>
+        private void onAsyncDelegateUnhandledException(Task task, object state)
+        {
+            AsyncTaskInfo delegateInfo;
+            lock (this.lockAsyncDelegates)
+            {
+                if (this.asyncDelegates.TryGetValue((IAsyncDelegate)state, out delegateInfo))
+                {
+                    // casted to IAsyncTaskInfoSetter to be able to set properties
+                    IAsyncTaskInfoSetter infoSetter = delegateInfo;
+
+                    infoSetter.Exception = task.Exception.GetBaseException();
+                    infoSetter.Status = task.Status;
+                }
+                else
+                {
+                    // Should never happen.
+                    this.logger.LogError("Cannot find the AsyncDelegateInfo related to the faulted task with Id {0}.", task.Id);
+                    return;
+                }
+
+                this.logger.LogError(task.Exception.GetBaseException(), "Unhandled exception for async delegate worker {0}.", delegateInfo.FriendlyName);
+            }
+        }
+
+        /// <summary>
+        ///  This method is called when a Task running an <see cref="IAsyncDelegate"/> completed or was canceled.
+        ///  It removes the task information from the internal dictionary.
+        /// </summary>
+        /// <param name="task">The delegate task.</param>
+        /// <param name="state">The <see cref="IAsyncDelegate"/> that's run by the delegateTask</param>
+        private void onAsyncDelegateCompleted(Task task, object state)
+        {
+            AsyncTaskInfo itemToRemove;
+            lock (this.lockAsyncDelegates)
+            {
+                if (this.asyncDelegates.TryGetValue((IAsyncDelegate)state, out itemToRemove))
+                {
+                    // When AsyncLoop fails with an uncaughtException, it handle it completing fine.
+                    // I want instead to keep its failed status visible on console so I handle this scenario as faulted task.
+                    // TODO: discuss about this decision.
+                    if (state is AsyncLoop asyncLoop && asyncLoop.UncaughtException != null)
+                    {
+                        // casted to IAsyncTaskInfoSetter to be able to set properties
+                        IAsyncTaskInfoSetter infoSetter = itemToRemove;
+
+                        infoSetter.Exception = asyncLoop.UncaughtException;
+                        infoSetter.Status = TaskStatus.Faulted;
+
+                        this.logger.LogTrace("Async Loop '{0}' completed with an UncaughtException, marking it as faulted. Task Id: {1}.", itemToRemove.FriendlyName, task.Id);
+                        return;
+                    }
+                    else
+                    {
+                        this.asyncDelegates.Remove((IAsyncDelegate)state);
```

Comment:
We should not remove anything from here right?

Reply:
here we are handling the task that completed succesfully, so we remove the delegate from the dictionary because it's not running anymore and it's an expected behavior from the creator of the loop/dequeuer

Basically I had to handle (like per comment above) a special case when a failing loop were eating an uncaughtException, otherwise this method had just to remove the entry from the dictionary

Reply:
ah ok thats the success path fine